src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs

```diff
@@ -0,0 +1,299 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal class WinHttpAuthHelper
+    {
+        // TODO:  This looks messy but it is fast. Research a cleaner way
+        // to do this which keeps high performance lookup.
+        //
+        // Fast lookup table to convert WINHTTP_AUTH constants to strings.
+        // WINHTTP_AUTH_SCHEME_BASIC = 0x00000001;
+        // WINHTTP_AUTH_SCHEME_DIGEST = 0x00000008;
+        // WINHTTP_AUTH_SCHEME_NEGOTIATE = 0x00000010;
+        private static readonly string[] s_authSchemeStringMapping =
+        {
+            null,
+            "Basic",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Digest",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Negotiate"
+        };
+
+        private static readonly uint[] s_authSchemePriorityOrder =
+        {
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_NEGOTIATE,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_DIGEST,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_BASIC
+        };
+
+        // TODO: This current design uses a handler-wide lock to Add/Retrieve
+        // from the cache.  Need to improve this for next iteration in order
+        // to boost performance and scalability.
+        private CredentialCache _credentialCache = new CredentialCache();
+        private object _credentialCacheLock = new object();
+        
+        public void CheckResponseForAuthentication(
+            WinHttpRequestState state,
+            ref uint proxyAuthScheme,
+            ref uint serverAuthScheme)
+        {
+            state.RetryRequest = false;
+
+            // Check the status code and retry the request applying credentials if needed.
+            var statusCode = (HttpStatusCode)WinHttpResponseParser.GetResponseHeaderNumberInfo(
+                state.RequestHandle,
+                Interop.WinHttp.WINHTTP_QUERY_STATUS_CODE);
+            uint supportedSchemes = 0;
+            uint firstSchemeIgnored = 0;
+            uint authTarget = 0;
+            Uri uri = state.RequestMessage.RequestUri;
+
+            switch (statusCode)
+            {
+                case HttpStatusCode.Unauthorized:
+                    if (state.ServerCredentials == null || state.LastStatusCode == HttpStatusCode.Unauthorized)
+                    {
+                        // Either we don't have server credentials or we already tried 
+                        // to set the credentials and it failed before.
+                        // So we will let the 401 be the final status code returned.
+                        break;
+                    }
```

Comment:
I thought this code block looked funny with the `break;` line hanging out in space which is why I had no newline between the `break` and the previous code block.

```c#
                    if (serverAuthScheme != 0)
                    {
                        SetWinHttpCredential(
                            state.RequestHandle,
                            state.ServerCredentials,
                            uri,
                            serverAuthScheme,
                            authTarget);

                        state.RetryRequest = true;
                    }
                    
                    break;

                case HttpStatusCode.ProxyAuthenticationRequired:
                    if (state.LastStatusCode == HttpStatusCode.ProxyAuthenticationRequired)
                    {
                        // We tried already to set the credentials.
                        break;
                    }
```